File: Problem3.py
#Time Complexity :O(n^2) 
#Space Complexity :O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        people.sort(key=lambda x:(-x[0],x[1]))
        output=[]
        for p in people:
            output.insert(p[1],p)
        return output

        

File: Problem_3.js
// ## Problem3 Partition Labels (https://leetcode.com/problems/partition-labels/)

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function (s) {
    if (s === null || s.length === 0)
        return [];
    let map = new Map();
    let result = [];
    // Map all the char and the last index they have occured
    for (let i = 0; i < s.length; i++) {
        map.set(s[i], i);
    }
    // mark start and end
    let start = 0;
    let end = map.get(s[0]);
    for (let i = 0; i < s.length; i++) {
        let ch = s[i];
        // Update end if the new character comes after the old character in string
        end = Math.max(end, map.get(ch));
        // If i === end, add the length to result
        if (i === end) {
            result.push(end - start + 1);
            start = i + 1;
        }
    }
    return result;
};
